File: php/src/Docxmerge.php
<?php

namespace Docxmerge;


class Docxmerge
{
    private $apiKey;
    private $baseUrl;
    private $tenant;

    public function __construct($apiKey, $baseUrl, $tenant = "default")
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = $baseUrl;
        $this->tenant = $tenant;
    }

    public function transformTemplate($fp, $templateName)
    {
        $baseUrl = $this->baseUrl;
        $ch = curl_init("$baseUrl/api/v1/Admin/TransformTemplate?template=$templateName");
        curl_setopt($ch, CURLOPT_FILE, $fp);
        $this->hydrateCurl($ch);

        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }

    public function transformDocument($fp, $docxPath)
    {
        $baseUrl = $this->baseUrl;
        $ch = curl_init("$baseUrl/api/v1/Admin/TransformFile");
        curl_setopt($ch, CURLOPT_FILE, $fp);
        $this->hydrateCurl($ch);
        $fields = [
            'file' => new \CurlFile($docxPath, '', 'file.docx'),
        ];
        curl_setopt($ch, CURLOPT_POSTFIELDS, $fields);
        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }

    public function mergeTemplate($fp, $templateName, $data)
    {
        $baseUrl = $this->baseUrl;
        $apiKey = $this->apiKey;
        $tenant = $this->tenant;

        $ch = curl_init("$baseUrl/api/v1/Admin/MergeTemplate?template=$templateName");

        $payload = json_encode($data);

        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
        curl_setopt($ch, CURLOPT_FILE, $fp);
        curl_setopt($ch, CURLOPT_POST, 1);
        $body_len = strlen($payload);
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                "Content-Length:$body_len",
                "api-key: $apiKey",
                "x-tenant: $tenant",
            )
        );
        curl_setopt($ch, CURLOPT_POSTFIELDS, $payload);

        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }

    public function mergeDocument($fp, $docxPath, $data)
    {
        $baseUrl = $this->baseUrl;
        $apiKey = $this->apiKey;
        $tenant = $this->tenant;

        $ch = curl_init("$baseUrl/api/v1/Admin/MergeFile");

        $payload = json_encode($data);

        $options = array(
            CURLOPT_POST => true,
            CURLOPT_HTTPHEADER => array(
                "api-key: $apiKey",
                "x-tenant: $tenant",
            ),
            CURLOPT_POSTFIELDS => array(
                'file' => new \CurlFile("$docxPath", '', 'file'),
                'data' => $payload,
            ),
            CURLOPT_FILE => $fp,
        );
        curl_setopt_array($ch, $options);
        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }

    public function mergeAndTransformTemplate($fp, $templateName, $data)
    {
        $baseUrl = $this->baseUrl;
        $apiKey = $this->apiKey;
        $tenant = $this->tenant;

        $ch = curl_init("$baseUrl/api/v1/Admin/MergeAndTransformTemplatePost?template=$templateName");

        $payload = json_encode($data);

        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
        curl_setopt($ch, CURLOPT_FILE, $fp);
        curl_setopt($ch, CURLOPT_POST, 1);
        $body_len = strlen($payload);
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                "Content-Length:$body_len",
                "api-key: $apiKey",
                "x-tenant: $tenant",
            )
        );
        curl_setopt($ch, CURLOPT_POSTFIELDS, $payload);

        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }

    public function mergeAndTransformDocument($fp, $docxPath, $data)
    {
        $baseUrl = $this->baseUrl;
        $apiKey = $this->apiKey;
        $tenant = $this->tenant;

        $ch = curl_init("$baseUrl/api/v1/Admin/MergeAndTransform");

        $payload = json_encode($data);

        $options = array(
            CURLOPT_POST => true,
            CURLOPT_HTTPHEADER => array(
                "api-key: $apiKey",
                "x-tenant: $tenant",
            ),
            CURLOPT_POSTFIELDS => array(
                'file' => new \CurlFile("$docxPath", '', 'file'),
                'data' => $payload,
            ),
            CURLOPT_FILE => $fp,
        );
        curl_setopt_array($ch, $options);
        curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($http_code != "200") throw new \Exception("Unexpected status code $http_code");
    }


    private function hydrateCurl($ch)
    {
        $apiKey = $this->apiKey;
        $tenant = $this->tenant;
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(
            "api-key: $apiKey",
            "x-tenant: $tenant",
        ));
    }

    /**
     * @return mixed
     */
    public function getBaseUrl()
    {
        return $this->baseUrl;
    }

    /**
     * @return mixed
     */
    public function getApiKey()
    {
        return $this->apiKey;
    }

    /**
     * @return string
     */
    public function getTenant(): string
    {
        return $this->tenant;
    }
}

File: ruby/lib/docxmerge.rb
require "docxmerge/version"
require 'net/https'
require 'uri'
require 'json'
require 'rest-client'

class BadRequestError < StandardError
end
module Docxmerge
  # Your code goes here...
  class Docxmerge
    def initialize(api_key, tenant, base_uri)
      @api_key = api_key
      @tenant = tenant || "default"
      @base_uri = base_uri || "https://api.docxmerge.com"
    end

    def render_template(template_name, data, conversion_type, version = "")
      uri = "#{@base_uri}/api/v1/Admin/RenderTemplate"
      headers = get_headers
      headers["content-type"] = "application/json"
      response = RestClient.post(uri,
                                 {
                                     template: template_name,
                                     data: data,
                                     version: version,
                                     conversionType: conversion_type,
                                 }.to_json,
                                 headers,
      )
      raise BadRequestError.new unless response.code == 200
      response.body
    end

    def render_file(file_param, data, conversion_type)
      uri = "#{@base_uri}/api/v1/Admin/RenderFile"
      response = RestClient.post(uri,
                                 {
                                     :multipart => true,
                                     :file => file_param,
                                     :data => data.to_json,
                                     :conversionType => conversion_type,
                                 },
                                 get_headers,
      )
      raise BadRequestError.new unless response.code == 200
      response.body
    end

    def render_url(url, data, conversion_type, version = "")
      uri = "#{@base_uri}/api/v1/Admin/RenderUrl"
      headers = get_headers
      headers["content-type"] = "application/json"
      response = RestClient.post(uri,
                                 {
                                     url: url,
                                     data: data,
                                     version: version,
                                     conversionType: conversion_type,
                                 }.to_json,
                                 headers,
      )
      raise BadRequestError.new unless response.code == 200
      response.body
    end

    private

    def get_headers
      {
          "api-key" => @api_key,
          "x-tenant" => @tenant,
      }
    end
  end
end
